router_types.c

```diff
@@ -44,6 +47,55 @@ static const struct router_type *router_types[] = {
 	NULL,
 };
 
+int read_mac(uint8_t mac[ETH_ALEN], const char *macstr)
+{
+	int ret;
+
+	ret = sscanf(macstr, "%02hhX:%02hhX:%02hhX:%02hhX:%02hhX:%02hhX",
+				 &mac[0], &mac[1], &mac[2], &mac[3], &mac[4], &mac[5]);
+	if (ret != 6)
+		ret = sscanf(macstr, "%02hhX-%02hhX-%02hhX-%02hhX-%02hhX-%02hhX",
+					 &mac[0], &mac[1], &mac[2], &mac[3], &mac[4], &mac[5]);
+
+	return (ret == 6);
+}
+
+void mac_whitelist_add(char *macstr)
+{
+	uint8_t mac[ETH_ALEN];
+	if (read_mac(mac, macstr)) {
+		if (mac_whitelist_head == NULL) {
+			mac_whitelist_head = malloc(sizeof(struct mac_whitelist_entry));
```

Comment:
Check for allocation errors. And use sizeof(*mac_whitelist_head)

Reply:
It is rather unlikely to fail here but just print an error for now and return from this function.

To make it even better, you could return -ENOMEM (change function return type to int and include errno.h). The non-error return value should be 0. The caller should check for values < 0 and handle this as an error (in your case stopping the program).